File: R/demEcoliMap1.R
rm(list = ls(all=TRUE))
source("gpsimLoadData.R")
source("miscFunctions.R")

source("kernFunctions.R")
source("mlpKernFunctions.R")
source("optimiFunctions.R")
source("gpsimMapFunctions.R")
source("translateKernFunctions.R")
source("cmpndKernFunctions.R")

expType <- "demEcoliMap"
expNo <- 1

options(error = recover)

trainingData <- gpsimLoadEcoliFullData()
y <- trainingData$y
genes <- trainingData$genes
times <- trainingData$times
scale <- trainingData$scale
rm(trainingData)

Nrep <- length(y)
Ngenes <- length(genes)
Ntf <- 1

yvar <- list()
for ( i in seq(length=Nrep) ) {
  yvar[[i]] <- matrix(0, dim(y[[i]])[1], dim(y[[i]])[2])
}

numGenes <- length(genes)


options <- list(includeNoise=1, optimiser="SCG")
# options$proteinPrior <- list(values=array(), times=array())
options$kern <- c("translate", "mlp")
options$nonLinearity <- "repression"
options$startPoint <- 0
options$endPoint <- 60
options$intPoints <- (options$endPoint-options$startPoint)/0.5+1
options$gParam <- matrix(1, 1, Ngenes) 
options$times <- times
options$bTransform <- "exp"
options$alphaTransform <- ""

options$S <- array(1, Ngenes)
options$D <- runif(Ngenes)
mu <- apply(y[[i]], 2, mean)
options$B <- options$D*mu

options0 <- options

model <- list(type="cgpsimMap")
for ( i in seq(length=Nrep) ) {
  display <- FALSE
  options <- gpsimMapInitParam(y[[i]], yvar[[i]], options0, display)
  if ( options$kern[2] == "mlp" ) {
    options$fix$index <- c(3)
    options$fix$value <- expTransform(c(1), "xtoa")
  } else if ( options$kern[2] == "rbf" ) {
    options$fix$index <- c(2, 19, 20, 21)
    options$fix$value <- expTransform(c(1, 1.75, 0.23, 0.01), "xtoa")
  }

  repNames <- names(model$comp)
  model$comp[[i]] <- gpsimMapCreate(Ngenes, Ntf, times, y[[i]], yvar[[i]], options)
  names(model$comp) <- c(repNames, paste("rep", i, sep=""))
  if ( options$kern[2] == "mlp" ) {
    model$comp[[i]]$kern$comp[[1]]$weightVariance <- 30
    model$comp[[i]]$kern$comp[[1]]$biasVariance <- 20
    model$comp[[i]]$kern$centre <- -15
    params <- gpsimMapExtractParam(model$comp[[i]])
    model$comp[[i]] <- gpsimMapExpandParam(model$comp[[i]], params)
  }
}

paramvec <- list()
param <- 0
for ( i in seq(length=Nrep) ) {
  options <- optimiFdefaultOptions()
  model$comp[[i]] <- gpsimMapUpdateF(model$comp[[i]], options)
  fmean <- mean(model$comp[[i]]$f)
  ypredMean <- apply(model$comp[[i]]$ypred, 2, mean)
  ypredScale <- sqrt(apply(model$comp[[i]]$ypred[model$comp[[i]]$timesIndex,], 2, var))
  model$comp[[i]]$B <- (model$comp[[i]]$B - model$comp[[i]]$D*(ypredMean - ypredScale*mean(model$comp[[i]]$y)))/ypredScale
    
  if ( any(grep("alpha", names(model))) ) 
    model$comp[[i]]$alpha <- model$comp[[i]]$alpha/ypredScale
   
  for ( j in seq(length=numGenes) ) {
    if ( model$comp[[i]]$B[j] < 0 ) {
      model$comp[[i]]$alpha[j] <- model$comp[[i]]$alpha[j] + model$comp[[i]]$B[j]
      model$comp[[i]]$B[j] <- 1e-6
    }
  }
  
  model$comp[[i]]$S <- model$comp[[i]]$S/ypredScale 
  f <- gpsimMapFunctionalExtractParam(model$comp[[i]])
  model$comp[[i]] <- gpsimMapFunctionalExpandParam(model$comp[[i]], f)
  
  paramvec[[i]] <- gpsimMapExtractParam(model$comp[[i]]) 
  param <- param + paramvec[[i]]    
}

param <- param/Nrep;
ll0 <- gpsimMapObjective(param, model)
iters <- 300;

optOptions <- optimiDefaultOptions()
optOptions$maxit <- 300
optOptions$optimiser <- "SCG"

optimResult <- SCGoptim(param, fn=gpsimMapObjective, grad=gpsimMapGradients, optOptions, model)

fileName <- paste(expType, expNo, ".Rdata", sep="")
save(model, expType, expNo, genes, scale, file=fileName)

for ( i in seq(length=Nrep) ) {
  optimiFoptions <- optimiFdefaultOptions()
  model$comp[[i]] <- gpsimMapExpandParam(model$comp[[i]], optimResult$xmin)
  model$comp[[i]] <- gpsimMapUpdateF(model$comp[[i]], optimiFoptions)
  model$comp[[i]] <- gpsimMapUpdateYpredVar(model$comp[[i]])
}

fileName <- paste(expType, expNo, ".Rdata", sep="")
save(model, expType, expNo, genes, scale, file=fileName)

displayOption <- 1
gpsimMapResults(model, expType, expNo, genes, scale, displayOption)

File: gpmgpsim/demEcoliMap1.m
% DEMECOLIMAP1 Optimise model using MAP approximation with MLP kernel and multiple repression response, using SCG optimisation and Ecoli data set.

% SHEFFIELDML

clear;
expNo = 4;
type = 'mapFullEcoliOptimInit';
rand('seed',2);

load demBarenco1;
origModel = model;
clear model;
clear y;
clear yvar;
saveFigures = false;

colordef white
[y, gene, times, scale, rawExp] = gpsimLoadEcoliFullData;
Nrep = length(y);
for i = 1:Nrep
  yvar{i} = zeros(size(y{i}));
end

numGenes = length(gene);
% Get the default options structure.
options = gpsimMapOptions(numGenes);
options.kern = {'translate', 'mlp'};    % 'mlp';
options.nonLinearity = 'repression';
options.includeNoise = 0;
options.startPoint = 0;                 % TF starts from 0;
options.endPoint = 60;
options.priorProteinTimes = []';
options.priorProtein = []';
options.includeNoise = 1;
options.bTransform = 'exp';
options.alphaTransform = [];
% Modify according to the range of the TF
options.intPoints = (options.endPoint-options.startPoint)/0.5 + 1 ;
options.gParam = ones(1,numGenes);
options.times = times;

for i =1:length(y)
    times = times;
    options = gpsimMapInitParam(y{i}, yvar{i}, options);
    
    if strcmp(options.kern{2}, 'mlp')
%       options.fix(1).index = 20;         % S of gene4
%       options.fix(1).value = expTransform(1.75, 'xtoa');
%       options.fix(2).index = 21;          % D of gene4
%       options.fix(2).value = expTransform(0.23, 'xtoa');
      options.fix(1).index = 3;  
      options.fix(1).value =  expTransform(1, 'xtoa');
%       options.fix(4).index = 22;  % alpha of gene 4
%       options.fix(4).value =  expTransform(0.01, 'xtoa');
    elseif strcmp(options.kern, 'rbf')
      options.fix(1).index = 19;         % S of gene4
      options.fix(1).value = expTransform(1.75, 'xtoa');
      options.fix(2).index = 20;          % D of gene4
      options.fix(2).value = expTransform(0.23, 'xtoa');
      options.fix(3).index = 2;  
      options.fix(3).value =  expTransform(1, 'xtoa');
      options.fix(4).index = 21;  % alpha of gene 4
      options.fix(4).value =  expTransform(0.01, 'xtoa');
    end        
    
    model.comp{i} = gpsimMapCreate(numGenes, 1, times, y{i}, yvar{i}, options);
    if strcmp(options.kern{2}, 'mlp')
        model.comp{i}.kern.comp{1}.weightVariance = 30;

        model.comp{i}.kern.comp{1}.biasVariance = 20;
        model.comp{i}.kern.centre = -15;
        % This forces kernel recompute.
        params = gpsimMapExtractParam(model.comp{i});
        model.comp{i} = gpsimMapExpandParam(model.comp{i}, params);
    elseif strcmp(options.kern, 'rbf')
      model.comp{i}.kern.inverseWidth = 0.01;
      params = gpsimMapExtractParam(model.comp{i});
      model.comp{i} = gpsimMapExpandParam(model.comp{i}, params);
    end
end

param = 0;

for i = 1:Nrep
  

    option = defaultOptions;
    model.comp{i} = gpsimMapUpdateF(model.comp{i}, option);
   
    % modify the initialisations
    fmean =  mean(model.comp{i}.f);
%/~
    %     fsd = sqrt(var(model.comp{i}.f));
%    model.comp{i}.f = model.comp{i}.f - fmean;
%     model.comp{i}.f = model.comp{i}.f/fsd;
%     model.comp{i}.S = model.comp{i}.S/exp(fmean);
%     model.comp{i}.gParam = model.comp{i}.gParam/exp(fmean);
%    ypredMean = mean(model.comp{i}.ypred);
%~/
    ypredScale = sqrt(var(model.comp{i}.ypred(model.comp{i}.times_index,:)));
    model.comp{i}.B = (model.comp{i}.B - model.comp{i}.D.* ...
                       (ypredMean - ypredScale.*mean(model.comp{i}.y)))./ypredScale;
    
    if isfield(model.comp{i}, 'alpha')
      model.comp{i}.alpha = model.comp{i}.alpha./ypredScale;
    end
    for j = 1:numGenes
      if model.comp{i}.B(j) < 0
        model.comp{i}.alpha(j) =  model.comp{i}.alpha(j) + model.comp{i}.B(j);
        model.comp{i}.B(j) = 1e-6;
      end
    end
    model.comp{i}.S = model.comp{i}.S./ypredScale; 
    f = gpsimMapFunctionalExtractParam(model.comp{i});
    model.comp{i} = gpsimMapFunctionalExpandParam(model.comp{i}, f);
    
    [paramvec{i} paramNames] = gpsimMapExtractParam(model.comp{i}); %vector of gamma estimates
    param = param + paramvec{i};    
    
end

param = param/Nrep;
ll0 = gpsimMapObjective(param, model)
iters = 300;

option = optOptions;
option(14) = iters;
option(9) = 0;
option(1) = 1;

newparam = scg('gpsimMapObjective', param,  option, ...
               'gpsimMapGradients', model);
%/~ 
% [newparam, ll, index] = minimize(param', 'gpsimMapGradFuncWrapper', iters, mo% del);
%~/
for rep = 1:Nrep
    option = defaultOptions; 
    model.comp{rep} = gpsimMapExpandParam(model.comp{rep}, newparam);   
    model.comp{rep} = gpsimMapUpdateF(model.comp{rep}, option);
    model.comp{rep}.yvar = ones(length(times), 1)* ...
        model.comp{rep}.noiseVar;
    model.comp{rep} = gpsimMapUpdateYpredVar(model.comp{rep});
end

type(1) = upper(type(1));  
save(['dem' type num2str(expNo)], 'model', 'type', 'expNo', 'scale')
gpsimMapEcoliResults(model, type, expNo, scale, saveFigures)
